apps/src/code-studio/pd/form_components/FormController.jsx

```diff
@@ -285,13 +290,34 @@ export default class FormController extends React.Component {
   validateCurrentPageRequiredFields() {
     const requiredFields = this.getRequiredFields();
     const pageFields = this.getCurrentPageComponent().associatedFields;
+
+    // Trim string values on page, and set empty strings to null
+    const pageData = {};
+    pageFields.forEach(field => {
+      let value = this.state.data[field];
+      if (typeof value === "string") {
+        const trimmedValue = value.trim();
+        pageData[field] = trimmedValue.length > 0 ? trimmedValue : null;
+      } else {
+        pageData[field] = value;
+      }
+    });
+    this.setState({
+      data: {
+        ...this.state.data,
+        ...pageData
+      }
+    });
+
```

Comment:
CC @Hamms - in order to fix a few of the issues, I improved the general form validation in a few places. PTAL. I'm also happy to pull these out into a separate PR if we think that's easier to review.

Improvements: 
- Trim all string (input, free text) values and set empty to null so just entering all spaces won't count as providing a required field.
- prevent default on the submit action, and validate the final page in client before actually submitting.
- When removing the last item in an array field (i.e. checkboxes), set it to null again so it won't count as providing a required field.

Reply:
This is fine for review. I love those improvements, but it's definitely worth checking the other forms to make sure none of them invalidate any assumptions; the empty array-to-null conversion in particular seems potentially dangerous.

I'd also love to see tests!

Reply:
👍  I'll add some tests, and look through the existing forms. 

I do however think the array logic is safe from playing around with it in this form. All fields at first are null, and will fail the required fields check. When a checkbox is selected, that field now [becomes an array](https://github.com/code-dot-org/code-dot-org/blob/dd15b20e3b798288fc227fbcc8fd9738b037277d/apps/src/code-studio/pd/form_components/button_list.jsx#L55-L61) containing the selected value. Subsequent selections are added, and de-selections are deleted. Before this change, whenever the last checkbox was unchecked, the field would still contain an empty array and pass the required field check. In other words, you could check, uncheck, and submit successfully even though the *effective* state of the form data does not contain a response for that field (the same as the initial state that failed validation).

Reply:
ok, I added tests for `FormController` and updated the ButtonList tests. PTAL